text/0000-grand-testing-unification.md

```diff
@@ -0,0 +1,647 @@
+
+In acceptance tests, we have two main API's for this today:
+
+- `Ember.Test.registerAsyncHelper`
+- `Ember.Test.registerHelper`
+
+These are run eagerly when your test suite is evaluated after which all acceptance tests have access to a global method with the provided helper name.  There are at least a few issues with this system:
+
+- The helpers receive the `Ember.Application` instance as the first argument.  This prevents us from utilizing the new `ApplicationInstance` instances, which would allow our acceptance testing to be much faster.
+- Having these helpers automatically augment the testing harness is convenient but often makes tracking down where a given helper is actually defined quite difficult.
+- It is also not obvious where you put this global side-effect registration code in a standard ember-cli application.
+- This does not work at all for unit or integration tests.
+
+In order to avoid these issues, I propose that test helpers should each be in their own module (`tests/helpers/*.js`) that is imported and registered on the test context:
+
+```js
+import moduleForAcceptance from '../helpers/module-for-acceptance';
+import { test } from 'ember-qunit';
+import selectCategory from '../helpers/select-category';
+import loginAsAdmin from '../helpers/login-as-admin';
+
+moduleForAcceptance('Can see things', {
+  beforeEach() {
+    this.registerHelpers({ loginAsAdmin, selectCategory});
+  }
+});
+
+test('stuff happens', function(assert) {
+  this.loginAsAdmin();
+  this.selectCategory('Islay');
+});
+```
+
+To author these custom helpers, you would do the following:
+
+```js
+// tests/helpers/login-as-admin';
+import { testHelper } from 'ember-qunit';
+
+export default testHelper(async function() {
+  await this.click('.login-now');
+  await this.fillIn('.username', 'rwjblue');
+  await this.fillIn('.password', 'moar beerz plz');
+  await this.click('.submit');
+});
+```
+
+- Due to the usage of `async` / `await` no special casing of sync vs async helpers is needed.
+- The helper shares a context with the test, so it can access the other helpers or properties as needed.  This is somewhat dependent on testing framework (some pass arguments for context while others like QUnit assume an implicit `this` context is provided by the framework).
+
+#### Registering Custom Waiters
+
+When the acceptance test helpers were initially released the system waited on three main things:
+
+- run loop queues to be flushed
+- all route async hooks to be completed (`beforeModel`, `model`, `afterModel`, etc)
+- pending AJAX requests
+
+It soon became clear that we needed to expose a way for users to influence what is waited on after each async interaction (i.e. `click`). Thus the `Ember.Test.registerWaiters` and `Ember.Test.unregisterWaiters` API's were added.  Unfortunately, these API's were global and users were forced through annoying hoops to use them in a modules world.
```

Comment:
Would be great if waiters could be attached to a certain async helper call instead of registering them globally. Possible public API (based on the example below):
```javascript
import transactionWaiter from '../waiters/pending-transactions';
...
test('stuff happens', async function(assert) {
  await this.click('.foo', transactionWaiter);
  // all pending transactions are completed here...
  await this.click('.bar');
  // second click doesn't care about transactions
});
```